linera-rpc/src/simple/client.rs

```diff
@@ -149,25 +149,39 @@ impl ValidatorNode for SimpleClient {
         &self,
         hash: CryptoHash,
     ) -> Result<ConfirmedBlockCertificate, NodeError> {
-        let response = self.download_certificates(vec![hash]).await?;
-        if response.len() != 1 {
-            Err(NodeError::MissingCertificates {
-                missing: hash.to_string(),
-            })
-        } else {
-            // UNWRAP: Safe b/c we just checked if len != 1.
-            Ok(response.into_iter().next().unwrap())
-        }
+        Ok(self
+            .download_certificates(vec![hash])
+            .await?
+            .into_iter()
+            .next()
+            .unwrap()) // UNWRAP: We know there is exactly one certificate, otherwise we would have an error.
     }
 
     async fn download_certificates(
         &self,
         hashes: Vec<CryptoHash>,
     ) -> Result<Vec<ConfirmedBlockCertificate>, NodeError> {
-        self.query::<Vec<ConfirmedBlockCertificate>>(RpcMessage::DownloadCertificates(Box::new(
-            hashes,
-        )))
-        .await
+        let certificates = self
+            .query::<Vec<ConfirmedBlockCertificate>>(RpcMessage::DownloadCertificates(Box::new(
+                hashes.clone(),
+            )))
+            .await?;
+
+        if certificates.len() != hashes.len() {
+            let missing_hashes: Vec<CryptoHash> = hashes
+                .into_iter()
+                .filter(|hash| !certificates.iter().any(|cert| cert.hash() == *hash))
+                .collect();
+            Err(NodeError::MissingCertificates {
+                missing: missing_hashes
+                    .iter()
+                    .map(|hash| hash.to_string())
+                    .collect::<Vec<_>>()
+                    .join(", "),
+            })
+        } else {
+            Ok(certificates)
+        }
```

Comment:
In the `grpc` version it's easier b/c we track `missing_hashes` already so we just check if it's non-empty.

Reply:
I don't think it's that simple. We'd have to:
- zip `hashes` with `certificates`
- take while corresponding elements' hashes are equivalent
- stop as soon as they're not.

This would tell us which certificates have been downloaded but we then need to _reverse_ that - b/c we want which certs are missing.

Is it worth the effort? This is the "pessimistic case" which shouldn't happen.
----

I rewrote to a difference on two hash sets.

Reply:
It depends. I imagine that at all our call sites we really only care about whether we got _all_ the certificates we asked for. Anyway, not a blocker!

Reply:
So we could downgrade the error to not even contain the missing hashes. If you're not feeling strong about it - I'd leave as-is for now.